doc/manual/metaprogramming.rst

```diff
@@ -9,106 +9,120 @@
-and a long form for blocks of code, enclosed in ``quote ... end``. Here
-is an example of the short form used to quote an arithmetic expression:
+Every Julia program starts life as a string:
 
 .. doctest::
 
-    julia> ex = :(a+b*c+1)
-    :(a + b * c + 1)
+    julia> prog = "1 + 1"
+    "1 + 1"
+
+**What happens next?**
+
+The next step is to `parse <http://en.wikipedia.org/wiki/Parsing#Computer_languages>`_
+each string into an object called an expression, represented by the Julia type
+:obj:`Expr`:
+
+.. doctest::
+
+    julia> ex1 = parse(prog)
+    :(1 + 1)
 
     julia> typeof(ex)
     Expr
 
-    julia> ex.head
+Expressions may also be constructed directly in
+`prefix notation <http://en.wikipedia.org/wiki/Polish_notation>`_:
+
+.. doctest::
+
+    julia> ex2 = Expr(:call, :+, 1, 1)
+    true
+
+The two expressions constructed above -- by parsing and by direct
+construction -- are equivalent:
+
+.. doctest::
+
+    julia> ex1 == ex2
+    true
+
+**The key point here is that Julia code is internally represented
+as a data structure that is accessible from the language itself.**
+
+:obj:`Expr` objects contain three parts:
+
+- a Symbol identifying the action represented by the expression:
+
+.. doctest::
+
+    julia> ex1.head
     :call
 
-    julia> typeof(ans)
-    Symbol
+- the expression arguments, which may be symbols, other expressions, or literal values:
+
+.. doctest::
 
-    julia> ex.args
-    4-element Array{Any,1}:
+    julia> ex1.args
+    3-element Array{Any,1}:
       :+
-      :a
-      :(b * c)
+     1
      1
 
-    julia> typeof(ex.args[1])
-    Symbol
+- finally, the expression result type, which may be annotated by the user or inferred
+  by the compiler (and may be ignored completely for the purposes of this chapter):
 
-    julia> typeof(ex.args[2])
-    Symbol
+.. doctest::
 
-    julia> typeof(ex.args[3])
-    Expr
+    julia> ex1.typ
+    Any
 
-    julia> typeof(ex.args[4])
-    Int64
+The :func:`dump` function provides indented and annotated display of :obj:`Expr`
+objects:
 
-Expressions provided by the parser generally only have symbols, other
-expressions, and literal values as their args, whereas expressions
-constructed by Julia code can easily have arbitrary run-time values
-without literal forms as args. In this specific example, ``+`` and ``a``
-are symbols, ``*(b,c)`` is a subexpression, and ``1`` is a literal
-64-bit signed integer. Here's an example of the longer expression
-quoting form:
+.. doctest::
+
+    julia> dump(ex2)
+    Expr
+      head: Symbol call
+      args: Array(Any,(3,))
+	1: Symbol +
+	2: Int64 1
+	3: Int64 1
+      typ: Any
+
+:obj:`Expr` objects may also be nested:
 
 .. doctest::
 
-    julia> quote
-             x = 1
-             y = 2
-             x + y
-           end
-    quote  # none, line 2:
-        x = 1 # line 3:
-        y = 2 # line 4:
-        x + y
-    end
+    julia> ex3 = parse("(4 + 4) / 2")
+    :((4 + 4) / 2)
+
+    julia> Base.Meta.show_sexpr(ex3)
```

Comment:
Since the topic is "display," one alternative phrasing might be something like:

Another way to view expressions is with `Meta.show_sexpr`, which displays the `S-expression <http://en.wikipedia.org/wiki/S-expression>`_ form of a given :obj:`Expr`, which may look very familiar to users of Lisp. Here's an example illustrating the display on a nested :obj:`Expr`: